torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -3136,6 +3147,17 @@ void ProcessGroupNCCL::assignTimeoutToWork(
 
 void ProcessGroupNCCL::workEnqueue(
     const c10::intrusive_ptr<ProcessGroupNCCL::WorkNCCL>& work) {
+  // We clean up the TensorShelf's in case user hasn't called `work.wait()`.
+  // This has nothing to do with new work enqueue. We are just using a place
+  // that would be triggered by a next user call.
+  {
+    std::lock_guard<std::mutex> lock(shelvesMutex_);
+    for (auto& shelf : shelvesToUnstash_) {
+      shelf->clear();
+    }
+    shelvesToUnstash_.clear();
```

Comment:
are we worried about extra memory allocations caused by long pauses between the work finishing and new work being added?

Reply:
I was a little concerned. But most of the time we are relying on the user to call `work.wait()` to clear the reference. The mechanism added here is more of a safety net in case the user didn't.

Reply:
It is not an actual move per se, but instead sharing a pointer to the *same* shelf that user's work is holding. So whichever call comes first, wait or enqueue, that shelf will be cleaned. Sorry about the "move" wording in my PR description.

Reply:
With previous design, there is a chance that watchdog does the cleaning earlier than user's `work.wait()`, if user didn't call it for long enough. This change kinda makes it less aggressive.